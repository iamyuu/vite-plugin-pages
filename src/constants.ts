export const MODULE_IDS = [
  '~react-pages',
  'virtual:generated-pages-react',
]

export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages'
export const ROUTE_IMPORT_NAME = '__pages_import_$1__'

export const dynamicRouteRE = /^\[(.+)\]$/
export const cacheAllRouteRE = /^\[\.{3}(.*)\]$/
export const replaceDynamicRouteRE = /^\[(?:\.{3})?(.*)\]$/

export const nuxtDynamicRouteRE = /^_(.*)$/
export const nuxtCacheAllRouteRE = /^_$/

export const countSlashRE = /\//g

export const replaceIndexRE = /\/?index$/
