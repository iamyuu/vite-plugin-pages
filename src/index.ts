import { MODULE_ID_VIRTUAL } from './constants'
import { PageContext } from './context'

import { parsePageRequest } from './utils'
import type { UserOptions } from './types'
import type { Plugin } from 'vite'

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let ctx: PageContext

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    async configResolved(config) {
      userOptions.resolver = 'react'

      ctx = new PageContext(userOptions, config.root)
      ctx.setLogger(config.logger)
      await ctx.searchGlob()
    },
    configureServer(server) {
      ctx.setupViteServer(server)
    },
    resolveId(id) {
      if (ctx.options.moduleIds.includes(id))
        return `${MODULE_ID_VIRTUAL}?id=${id}`

      return null
    },
    async load(id) {
      const {
        moduleId,
        pageId,
      } = parsePageRequest(id)

      if (moduleId === MODULE_ID_VIRTUAL && pageId && ctx.options.moduleIds.includes(pageId))
        return ctx.resolveRoutes()

      return null
    },
  }
}

export * from './types'
export type { ReactRoute } from './resolvers'

export { syncIndexResolver } from './options'
export { PageContext }
export default pagesPlugin
