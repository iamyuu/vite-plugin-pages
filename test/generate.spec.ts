import { describe, expect, test } from 'vitest'
import { PageContext } from '../src/context'

const sensitivity = process.platform === 'win32' ? 'base' : 'variant'

function deepSortArray(arr: any[]) {
  const key = 'element'
  arr.forEach((i) => {
    if (i.children)
      i.children = deepSortArray(i.children)
  })
  if (arr.length === 1) return arr
  return arr.sort((a, b) => {
    return a[key] && b[key] ? a[key].localeCompare(b[key], 'en', { sensitivity }) : !(a[key] && b[key]) ? 0 : a[key] ? 1 : -1
  })
}

describe('Generate routes', () => {
  test('react - match snapshot', async() => {
    const ctx = new PageContext({
      dirs: 'examples/next-style/src/pages',
      resolver: 'react',
      onRoutesGenerated(routes) {
        // eslint-disable-next-line no-console
        routes = deepSortArray(routes)
        expect(routes).toMatchSnapshot('routes')
        return routes
      },
    })
    await ctx.searchGlob()
    const routes = await ctx.resolveRoutes()

    expect(routes).toMatchSnapshot('client code')
  })

  describe('routeStyle', () => {
    test('nuxt style match snapshot', async() => {
      const ctx = new PageContext({
        dirs: 'examples/nuxt-style/src/pages',
        routeStyle: 'nuxt',
        resolver: 'react',
        onRoutesGenerated(routes) {
        // eslint-disable-next-line no-console
          routes = deepSortArray(routes)
          expect(routes).toMatchSnapshot('routes')
          return routes
        },
      })
      await ctx.searchGlob()
      const routes = await ctx.resolveRoutes()

      expect(routes).toMatchSnapshot('client code')
    })

    test('remix Style match snapshot', async() => {
      const ctx = new PageContext({
        dirs: 'examples/remix-style/src/pages',
        routeStyle: 'remix',
        resolver: 'react',
        onRoutesGenerated(routes) {
        // eslint-disable-next-line no-console
          routes = deepSortArray(routes)
          expect(routes).toMatchSnapshot('routes')
          return routes
        },
      })
      await ctx.searchGlob()
      const routes = await ctx.resolveRoutes()

      expect(routes).toMatchSnapshot('client code')
    })
  })
})
