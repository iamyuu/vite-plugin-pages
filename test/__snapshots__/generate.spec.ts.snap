// Vitest Snapshot v1

exports[`Generate routes > react - match snapshot > client code 1`] = `
"import React from \\"react\\";
const __pages_import_0__ = React.lazy(() => import(\\"/examples/next-style/src/pages/[...all].tsx\\"));
const __pages_import_1__ = React.lazy(() => import(\\"/examples/next-style/src/pages/about.tsx\\"));
const __pages_import_2__ = React.lazy(() => import(\\"/examples/next-style/src/pages/about/index.tsx\\"));
const __pages_import_3__ = React.lazy(() => import(\\"/examples/next-style/src/pages/components.tsx\\"));
import __pages_import_4__ from \\"/examples/next-style/src/pages/index.tsx\\";
const __pages_import_5__ = React.lazy(() => import(\\"/examples/next-style/src/pages/blog/[...all].tsx\\"));
const __pages_import_6__ = React.lazy(() => import(\\"/examples/next-style/src/pages/blog/[id].tsx\\"));
const __pages_import_7__ = React.lazy(() => import(\\"/examples/next-style/src/pages/blog/index.tsx\\"));
const __pages_import_8__ = React.lazy(() => import(\\"/examples/next-style/src/pages/blog/today/[...all].tsx\\"));
const __pages_import_9__ = React.lazy(() => import(\\"/examples/next-style/src/pages/blog/today/index.tsx\\"));

const routes = [{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_0__)},{\\"caseSensitive\\":false,\\"path\\":\\"about\\",\\"element\\":React.createElement(__pages_import_1__),\\"children\\":[{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_2__),\\"index\\":true}]},{\\"caseSensitive\\":false,\\"path\\":\\"components\\",\\"element\\":React.createElement(__pages_import_3__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_4__),\\"index\\":true},{\\"caseSensitive\\":false,\\"path\\":\\"blog\\",\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_5__)},{\\"caseSensitive\\":false,\\"path\\":\\":id\\",\\"element\\":React.createElement(__pages_import_6__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_7__),\\"index\\":true},{\\"caseSensitive\\":false,\\"path\\":\\"today\\",\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_8__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_9__),\\"index\\":true}]}]}];

export default routes;"
`;

exports[`Generate routes > react - match snapshot > routes 1`] = `
[
  {
    "caseSensitive": false,
    "element": "/examples/next-style/src/pages/[...all].tsx",
    "path": "*",
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/next-style/src/pages/about/index.tsx",
        "index": true,
      },
    ],
    "element": "/examples/next-style/src/pages/about.tsx",
    "path": "about",
  },
  {
    "caseSensitive": false,
    "element": "/examples/next-style/src/pages/components.tsx",
    "path": "components",
  },
  {
    "caseSensitive": false,
    "element": "/examples/next-style/src/pages/index.tsx",
    "index": true,
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/next-style/src/pages/blog/[...all].tsx",
        "path": "*",
      },
      {
        "caseSensitive": false,
        "element": "/examples/next-style/src/pages/blog/[id].tsx",
        "path": ":id",
      },
      {
        "caseSensitive": false,
        "element": "/examples/next-style/src/pages/blog/index.tsx",
        "index": true,
      },
      {
        "caseSensitive": false,
        "children": [
          {
            "caseSensitive": false,
            "element": "/examples/next-style/src/pages/blog/today/[...all].tsx",
            "path": "*",
          },
          {
            "caseSensitive": false,
            "element": "/examples/next-style/src/pages/blog/today/index.tsx",
            "index": true,
          },
        ],
        "path": "today",
      },
    ],
    "path": "blog",
  },
]
`;

exports[`Generate routes > routeStyle > nuxt style match snapshot > client code 1`] = `
"import React from \\"react\\";
const __pages_import_0__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/_.tsx\\"));
const __pages_import_1__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/about.tsx\\"));
const __pages_import_2__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/about/index.tsx\\"));
const __pages_import_3__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/components.tsx\\"));
import __pages_import_4__ from \\"/examples/nuxt-style/src/pages/index.tsx\\";
const __pages_import_5__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/blog/_.tsx\\"));
const __pages_import_6__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/blog/_id.tsx\\"));
const __pages_import_7__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/blog/index.tsx\\"));
const __pages_import_8__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/blog/today/_.tsx\\"));
const __pages_import_9__ = React.lazy(() => import(\\"/examples/nuxt-style/src/pages/blog/today/index.tsx\\"));

const routes = [{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_0__)},{\\"caseSensitive\\":false,\\"path\\":\\"about\\",\\"element\\":React.createElement(__pages_import_1__),\\"children\\":[{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_2__),\\"index\\":true}]},{\\"caseSensitive\\":false,\\"path\\":\\"components\\",\\"element\\":React.createElement(__pages_import_3__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_4__),\\"index\\":true},{\\"caseSensitive\\":false,\\"path\\":\\"blog\\",\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_5__)},{\\"caseSensitive\\":false,\\"path\\":\\":id\\",\\"element\\":React.createElement(__pages_import_6__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_7__),\\"index\\":true},{\\"caseSensitive\\":false,\\"path\\":\\"today\\",\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_8__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_9__),\\"index\\":true}]}]}];

export default routes;"
`;

exports[`Generate routes > routeStyle > nuxt style match snapshot > routes 1`] = `
[
  {
    "caseSensitive": false,
    "element": "/examples/nuxt-style/src/pages/_.tsx",
    "path": "*",
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/nuxt-style/src/pages/about/index.tsx",
        "index": true,
      },
    ],
    "element": "/examples/nuxt-style/src/pages/about.tsx",
    "path": "about",
  },
  {
    "caseSensitive": false,
    "element": "/examples/nuxt-style/src/pages/components.tsx",
    "path": "components",
  },
  {
    "caseSensitive": false,
    "element": "/examples/nuxt-style/src/pages/index.tsx",
    "index": true,
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/nuxt-style/src/pages/blog/_.tsx",
        "path": "*",
      },
      {
        "caseSensitive": false,
        "element": "/examples/nuxt-style/src/pages/blog/_id.tsx",
        "path": ":id",
      },
      {
        "caseSensitive": false,
        "element": "/examples/nuxt-style/src/pages/blog/index.tsx",
        "index": true,
      },
      {
        "caseSensitive": false,
        "children": [
          {
            "caseSensitive": false,
            "element": "/examples/nuxt-style/src/pages/blog/today/_.tsx",
            "path": "*",
          },
          {
            "caseSensitive": false,
            "element": "/examples/nuxt-style/src/pages/blog/today/index.tsx",
            "index": true,
          },
        ],
        "path": "today",
      },
    ],
    "path": "blog",
  },
]
`;

exports[`Generate routes > routeStyle > remix Style match snapshot > client code 1`] = `
"import React from \\"react\\";
const __pages_import_0__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/__marketing.tsx\\"));
const __pages_import_1__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/__marketing/product.tsx\\"));
const __pages_import_2__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/[sitemap.xml].tsx\\"));
const __pages_import_3__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/$.tsx\\"));
const __pages_import_4__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/about.tsx\\"));
const __pages_import_5__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/about/index.tsx\\"));
const __pages_import_6__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog.authors.tsx\\"));
const __pages_import_7__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog.tsx\\"));
const __pages_import_8__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog/$.tsx\\"));
const __pages_import_9__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog/$id.tsx\\"));
const __pages_import_10__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog/index.tsx\\"));
const __pages_import_11__ = React.lazy(() => import(\\"/examples/remix-style/src/pages/blog/today/index.tsx\\"));
import __pages_import_12__ from \\"/examples/remix-style/src/pages/index.tsx\\";

const routes = [{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_0__),\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"product\\",\\"element\\":React.createElement(__pages_import_1__)}]},{\\"caseSensitive\\":false,\\"path\\":\\"sitemap.xml\\",\\"element\\":React.createElement(__pages_import_2__)},{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_3__)},{\\"caseSensitive\\":false,\\"path\\":\\"about\\",\\"element\\":React.createElement(__pages_import_4__),\\"children\\":[{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_5__),\\"index\\":true}]},{\\"caseSensitive\\":false,\\"path\\":\\"blog/authors\\",\\"element\\":React.createElement(__pages_import_6__)},{\\"caseSensitive\\":false,\\"path\\":\\"blog\\",\\"element\\":React.createElement(__pages_import_7__),\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_8__)},{\\"caseSensitive\\":false,\\"path\\":\\":id\\",\\"element\\":React.createElement(__pages_import_9__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_10__),\\"index\\":true},{\\"caseSensitive\\":false,\\"path\\":\\"today\\",\\"children\\":[{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_11__),\\"index\\":true}]}]},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_12__),\\"index\\":true}];

export default routes;"
`;

exports[`Generate routes > routeStyle > remix Style match snapshot > routes 1`] = `
[
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/remix-style/src/pages/__marketing/product.tsx",
        "path": "product",
      },
    ],
    "element": "/examples/remix-style/src/pages/__marketing.tsx",
    "path": undefined,
  },
  {
    "caseSensitive": false,
    "element": "/examples/remix-style/src/pages/[sitemap.xml].tsx",
    "path": "sitemap.xml",
  },
  {
    "caseSensitive": false,
    "element": "/examples/remix-style/src/pages/$.tsx",
    "path": "*",
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/remix-style/src/pages/about/index.tsx",
        "index": true,
      },
    ],
    "element": "/examples/remix-style/src/pages/about.tsx",
    "path": "about",
  },
  {
    "caseSensitive": false,
    "element": "/examples/remix-style/src/pages/blog.authors.tsx",
    "path": "blog/authors",
  },
  {
    "caseSensitive": false,
    "children": [
      {
        "caseSensitive": false,
        "element": "/examples/remix-style/src/pages/blog/$.tsx",
        "path": "*",
      },
      {
        "caseSensitive": false,
        "element": "/examples/remix-style/src/pages/blog/$id.tsx",
        "path": ":id",
      },
      {
        "caseSensitive": false,
        "element": "/examples/remix-style/src/pages/blog/index.tsx",
        "index": true,
      },
      {
        "caseSensitive": false,
        "children": [
          {
            "caseSensitive": false,
            "element": "/examples/remix-style/src/pages/blog/today/index.tsx",
            "index": true,
          },
        ],
        "path": "today",
      },
    ],
    "element": "/examples/remix-style/src/pages/blog.tsx",
    "path": "blog",
  },
  {
    "caseSensitive": false,
    "element": "/examples/remix-style/src/pages/index.tsx",
    "index": true,
  },
]
`;
